docs/halibut-protocol.md

```diff
@@ -0,0 +1,83 @@
+# Halibut Protocol data exchange
+
+This document outlines the data that is exchanged between the **Client** and **Service**. In this document:
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus Server.
+ - **Service** is what executes the methods e.g. Tentacle. 
+
+
+When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.
+
+ ### Stage 1 - Identification
+ 
+  A TCP connection is established and both Service and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
+
+### Stage 2 - Message exchange
+
+Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:
+
+1. The Client sends a `Request` message containing the method to execute and data.
+2. The Service executes the method and sends the result in a `Response` message.
```

Comment:
Similar to above, "The Service executes the method and sends a `Response` message containing the result and 0 or more data streams`